### PR TITLE
Modify `BaseDevice` signature for channels

### PR DESCRIPTION
**Pull request by HGSilveri**
Originally, device declaration was not meant to be user-facing, which allowed for the device channels to be given under the `_channels` keyword in a weird format (a tuple of (channel ID, channel object) pairs). 

However, now that users should be able to create their custom devices, this becomes a problem. 

This PR changes the signature for attributing devices to a sequence in a few important ways:
- It splits the channel IDs (`channel_ids`) from the channel objects (`channel_objects`)
- Deprecates the `_channels` argument (to be removed in v0.10)
- Makes the definition of channel IDs optional, relying on a default ID when a custom value is not provided

### CODE REVIEW

**Review by a-corni — COMMENTED** [review]
Everything Looks good to me, just have a look to the comments in pulser-core/pulser/channels/channels.py and pulser-core/pulser/devices/_device_datacls.py and I will approve.

File: pulser-core/pulser/devices/_devices.py [hunk 1]
```diff
@@ -26,42 +26,33 @@
...
-                min_duration=16,
-                max_duration=2**26,
-            ),
+    channel_objects=(
+        Rydberg.Global(
+            max_abs_detuning=2 * np.pi * 20,
+            max_amp=2 * np.pi * 2.5,
+            clock_period=4,
+            min_duration=16,
+            max_duration=2**26,
```
[the remark this answers is not in the document]
Comment (HGSilveri) [reply pulser-core/pulser/devices/_devices.py:35]: This is a style convention: whenever you are splitting the arguments in multiple lines, each one will have a comma after it. You will even find that black enforces this: if you try to delete the comma, black will just put it back.

File: pulser-core/pulser/devices/_device_datacls.py [hunk 2]
```diff
@@ -154,13 +152,83 @@ def type_check(
...
         def to_tuple(obj: tuple | list) -> tuple:
             if isinstance(obj, (tuple, list)):
                 obj = tuple(to_tuple(el) for el in obj)
             return obj
 
         # Turns mutable lists into immutable tuples
-        object.__setattr__(self, "_channels", to_tuple(self._channels))
+        for param in self._params():
+            if "channel" in param:
+                object.__setattr__(self, param, to_tuple(getattr(self, param)))
```
[the remark this answers is not in the document]
Comment (HGSilveri) [reply pulser-core/pulser/devices/_device_datacls.py:231]: Yes, exactly. We need them to be tuples in order for the object to be hashable

File: pulser-core/pulser/channels/channels.py [hunk 3]
```diff
@@ -78,3 +78,7 @@ class Microwave(Channel):
     def basis(self) -> Literal["XY"]:
         """The addressed basis name."""
         return "XY"
+
+    def default_id(self) -> str:
+        """Generates the default ID for indexing this channel in a Device."""
+        return f"mw_{self.addressing.lower()}"
```
[the remark this answers is not in the document]
Comment (HGSilveri) [reply pulser-core/pulser/channels/channels.py:84]: Well, although that would work, it goes against the spirit of subclassing, where the parent class holds what is transversal to all while the child classes define what is specific to them. The fact that one subclass wants to define their default ID different shouldn't affect its parent class. 
Also, what would happen if we wanted to add an extra `Channel` subclass? In your approach, we would have to go and change the parent class again.

File: tutorials/advanced_features/Output Modulation and EOM Mode.ipynb [hunk 4]
```diff
@@ -147,10 +147,7 @@
     "    name=\"test_device\",\n",
     "    dimensions=2,\n",
     "    rydberg_level=60,\n",
-    "    _channels=(\n",
-    "        (\"rydberg_global\", rydberg_ch),\n",
-    "        (\"raman_local\", raman_ch),\n",
-    "    ),\n",
+    "    channel_objects=(rydberg_ch, raman_ch),\n",
```
[the remark this answers is not in the document]
Comment (HGSilveri) [reply tutorials/advanced_features/Output Modulation and EOM Mode.ipynb:150]: For sure. I think the new tutorial is the best place, though

File: pulser-core/pulser/devices/_devices.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (a-corni) [reply pulser-core/pulser/devices/_devices.py:35]: Great, I did not know about this ! I guessed it was something like this as I noticed it happened regularly in the code.

File: pulser-core/pulser/devices/_device_datacls.py [hunk 5]
```diff
@@ -364,11 +432,13 @@ def _to_dict(self) -> dict[str, Any]:
 
     @abstractmethod
     def _to_abstract_repr(self) -> dict[str, Any]:
-        params = self._params()
+        ex_params = ("_channels", "channel_objects", "channel_ids")
+        params = {
+            k: v for k, v in self._params().items() if k not in ex_params
+        }
```
[the remark this answers is not in the document]
Comment (HGSilveri) [reply pulser-core/pulser/devices/_device_datacls.py:438]: Although performance is not a concern here, I'm okay with changing

File: tutorials/advanced_features/Output Modulation and EOM Mode.ipynb (hunk 4 above)
[the remark this answers is not in the document]
Comment (a-corni) [reply tutorials/advanced_features/Output Modulation and EOM Mode.ipynb:150]: Yes I agree, I already know how to introduce it in the new tutorial.

File: pulser-core/pulser/channels/channels.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (a-corni) [reply pulser-core/pulser/channels/channels.py:84]: Okay I get the spirit. Eventually the dict solution is closer to this, but I like it less as you have to define all the names in a dict. In this case it's all fine by me.

File: pulser-core/pulser/devices/_device_datacls.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (a-corni) [reply pulser-core/pulser/devices/_device_datacls.py:231]: Okay, I will keep this solution in mind.

File: pulser-core/pulser/devices/_device_datacls.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (a-corni) [reply pulser-core/pulser/devices/_device_datacls.py]: Great, I will now approve the review !

**Review by a-corni — APPROVED** [review]
Everything looks fine to me !

**Pull request merged**